### DISCUSSION TITLE
Dependent type systems

### DISCUSSION THREAD

**Discussion by dabrahams** [hylo-lang/hylo#1126] [Language design]
@GunpowderGuy [raised](https://github.com/orgs/hylo-lang/discussions/915#discussioncomment-7398315) the topic of dependent type systems, suggesting that [Idris2](https://github.com/idris-lang/Idris2) had some innovations that make dependent types easier to use, and that automated proof systems could be easily built upon them.  They offered to elaborate if we opened a thread for the topic, so here we are.

**Comment by GunpowderGuy** [comment]
Sorry for the late ( and incomplete ) reply, i am currently in a horrible time crunch. 

In the previous thread you asked why i mentioned dependent types under meta-programming. They are a kind of type system that allows meta-programming to generate generic programs in the traditional sense. But also programs that are generic across invariants you define with the type system. 

lets see a classic example 

```cpp
#include <iostream>
#include <string>
#include <type_traits>

// Primary template
template<bool B>
struct ExampleFunction;

// Specialization for true
template<>
struct ExampleFunction<true> {
    using type = int;  // Type A for true
    static type getValue() {
        return 42;
    }
};

// Specialization for false
template<>
struct ExampleFunction<false> {
    using type = std::string;  // Type B for false
    static type getValue() {
        return "Hello, world!";
    }
};

int main() {
    ExampleFunction<true>::type valueTrue = ExampleFunction<true>::getValue();
    ExampleFunction<false>::type valueFalse = ExampleFunction<false>::getValue();

    std::cout << "Value (true): " << valueTrue << std::endl;
    std::cout << "Value (false): " << valueFalse << std::endl;

    return 0;
}
```
and in idris2

```haskell
module Main

%default total

boolDependentFunction : (b : Bool) -> if b then Int else String
boolDependentFunction True = 42
boolDependentFunction False = "Hello, world!"

main : IO ()
main = do
  printLn (boolDependentFunction True)
  printLn (boolDependentFunction False)
```

As for user defined invariants. In a language like c++ you can define types that depend on types or values ( like std array ). However dependent type systems bridge gaps for you allowing things like a pair of arrays with size unknown at compile time, but of provable equal value.  Or functions with an output smaller than the input ( which can be used to prove the totality of functions ) 

https://github.com/stefan-hoeck/idris2-tutorial/blob/main/src/Tutorial/Dependent.md

This not only enables extra security but also supercharges meta-programming

**Reply by kyouko-taiga** [reply]
In my experience, functional correctness is a noble but difficult path. There are easy things to prove, like simple induction, and then there are while loops computing fixed points over work lists. AFAICT, dependent type systems put you on that difficult path because you can't easily use provably correct functions without buying into the whole "let's prove everything" mindset.

We have already thought of some form of value constraints for Hylo, which will likely cover things like requiring two arrays to provably have the same size. But full value-dependency, like proving that a function returns a value smaller than its input is not on the table at this point.

We'd need to be convinced that adopting dependent-types enable these use cases _without_ significantly increasing the complexity of mundane code. If using a function `T { some_predicate } -> U` requires to jump through too many hoops when we have an arbitrary `T`, then I'd say the complexity costs are too high. That's because the "supercharged" functions will be unusable from the simpler ones.

Your example is a little too small to determine whether my concerns are baseless.

**Reply by hosseinhaeri** [reply]
Sorry to poke my nose into this. I was invited by @GunpowderGuy via my thread on the Idris Discord. I hope that means I am not uninvited.

> AFAICT, dependent type systems put you on that difficult path because you can't easily use provably correct functions without buying into the whole "let's prove everything" mindset.

I'm afraid that's totally wrong. C++TMP is a dependently typed language. Surprise! :D And, who in C++TMP proves everything? Functional programming, on the other hand, isn't always practised for proving functional correctness. Most Idris programs, for example, don't go into Coq, Isabelle, etc theorem proving; or, even that of Agda, which is again dependently typed. Idris is meant to be a systems programming language.

> Your example is a little too small to determine whether my concerns are baseless.

I can confirm that his example is just a translation of the first example in the old Idris book into C++TMP. I have taught that book a number of times and I was never convinced that I should translate that book into C++TMP when I'm teaching the latter. The resulting C++TMP will become unorthodox and even sometimes antipatterned. What I'd rather do is that I show the equivalent Idris to my students when I'm teaching C++TMP and that helps them a lot to see the intention out of the huge syntactic noise.

Having said all that, I hadn't heard about Hylo until an hour ago. So, your cost-benefit analysis remains internal to the team.

**Reply by hosseinhaeri** [reply]
BTW, if you're interested in real C++TMP that gets help from Idris, see my teaching here:
https://www.youtube.com/watch?v=Q8R7QJkSifk&list=PL19CfOAWDwXq14FAZ4tmtvg4H_Ufy4m5y&pp=gAQBiAQB
My examples are not contrived there. With my students, we develop a compile-time list library. (A C++23 version of Andrei's macro-metaprogramming lists in [Modern C++ Design](https://www.amazon.co.uk/Modern-Design-Programming-Patterns-Depth-ebook/dp/B00AU3JUHG/ref=sr_1_1?crid=S89W6937U4U6&keywords=andrei+alexandrescu&qid=1700441167&sprefix=Andrei+Ale%2Caps%2C89&sr=8-1).)

**Reply by kyouko-taiga** [reply]
Thanks for giving your lights. You are very welcome here!

I talked about the road to functional correctness because @GunpowderGuy seems quite interested in that topic and because dependently type system are a prime target for people with similar interests, at least in my experience.

I do not think the comparison with C++ templates is completely fair, because we're aiming at definition-site checking rather than use-site checking in Hylo. IIUC, Idris uses a definition-site understanding of type checking too. Template expansion makes definition-site checking impractical and the complexity to go from use-site-checking to definition-checking is what scares me, not only in terms of implementation but also (and most importantly) in terms of the user model.

What I'm hoping to get from this thread is a couple of examples that:
1. show real use cases motivating Idris-like dependently-typed code, ideally for something other than functional correctness;
2. show that Idris-like dependent types do not significantly increase the complexity of non-dependently-typed code;
3. show that interacting between dependently and non-dependently-typed code is reasonably possible.

**Reply by dabrahams** [reply]
Separately, I do indeed find the idea that C++TMP is a dependently typed language quite surprising.  Its static type system is incredibly weak, having only a few types (type, various integers, various template and function pointer/reference shapes, …), and the only way to do anything resembling generic programming with it (e.g. treat lists of integers the same way as lists of types) is to erase these distinctions behind type wrappers, which basically defers _all_ type checking until TMP runtime (i.e. when templates are instantiated).  That makes it fundamentally dynamically typed.  For these reasons and others, there's almost nothing about the type system of C++ TMP that I'd want to replicate in Hylo.

I don't want to give an impression of hostility—I'm still interested in hearing about dependent typing—but since I have a fairly deep C++TMP background and my view is as above, I don't think it's going to make a very useful reference point in our discussions.

I'd be open to the idea that C++ itself has a dependent type system, but I don't think(?) the ability to encode numbers into generic parameters (which we do intend to support and you seem to center your video on) is—on its own—sufficient for calling a language dependently-typed.

**Reply by hosseinhaeri** [reply]
@kyouko-taiga 
Thank you for your warm welcome.
> Template expansion makes definition-site checking impractical

That's pre-C++20. C++ Concepts were such a game-changer. Modern C++TMP is only lenient to that but not such itself.

> the complexity to go from use-site-checking to definition-checking is what scares me, not only in terms of implementation but also (and most importantly) in terms of the user model.

I can't be the judge of your frights. But, I can say that the transition from the user's point of view is actually quite loveable. I used to teach C++TMP (with the list library example) according to metafunctions and all that pre-C++20. It was so loveable to move to C++23 with Concept-correctness, abbreviated function templates, and `consteval`.

> show real use cases motivating Idris-like dependently-typed code, ideally for something other than functional correctness;

Again, I'm not sure how motivating this might be to you because I don't know your inspirations (and those of Hylo). But, here is a trivial example:

```c++
template<auto...> List {};

template<auto X, auto... Xs> constexpr auto prepend(List<Xs...> l) {return List<X, Xs...>();}
``` 

Take your time please to think about the return type of `prepend`. How would have you used concepts to enforce that the length of the return List needs to be one more than that of `l`? Please note that the following would fail

```c++
template<auto... Xs> constexpr auto length(List<Xs...>) {return sizeof...(Xs);}

template<auto X, auto... Xs> constexpr auto prepend(List<Xs...> l)
requires (length(prepend(l)) == sizeof...(Xs) + 1) {return List<X, Xs...>();}
``` 
because, in C++, there are only two things concepts can do:

1. They help with the stage of error reporting and the quality of that.
2. They guide the compiler with overload resolution.

As I detail in my videos, the C++TMP type system is strictly less expressive than that of Idris. And, that's why they can't handle the above one-more-ness. That's bread-and-butter for Idris: (Gosh! Neither have Idris nor Haskell in the code-blocks of this forum...)

```
prepend : t -> Vect n t -> Vect (n + 1) t
prepend x l = x :: l
``` 

Why would one want to have that enforced? Correctness for sure. But, not just. That gives the compiler much more possibility for optimisation -- both to `prepend` and its clients.

> show that Idris-like dependent types do not significantly increase the complexity of non-dependently-typed code;

How can it? What in `4 + 5`, `string("this") + "that"`, `3.14 * 2.17` below would change at all in the presence of dependent-typing or its absence?

```c++
prepend<2 + 3>(List<4 + 5, string("this") + "that", 3.14 * 2.17>);
```

> show that interacting between dependently and non-dependently-typed code is reasonably possible.

How about the above? See my recorded videos for many more (real-world) examples.

**Reply by hosseinhaeri** [reply]
@dabrahams 
I read your book during my bachelors and loved it. Thanks a lot for the book. It's an honour to have you comment on my stuff. :)

> That makes it fundamentally dynamically typed.

Let me first remind that the stage at which types are processed has nothing to do with dependent-typing. In theory, a dependently-typed language can also be dynamically typed. (Is there such a language though?) Those are orthogonal topics.

> I don't think(?) the ability to encode numbers into generic parameters (which we do intend to support and you seem to center your video on) is—on its own—sufficient for calling a language dependently-typed.

A language is dependently-typed as soon as types and (otherwise) values can be passed around alike. A `consteval` function, for example, can return a non-type in as much as it can return a value of a given type so to cater `decltype`ing -- by doing which, effectively returning types (as done at several occasions during my tutorials).

Now, lets get to your perception of C++TMP...

> Its static type system is incredibly weak, having only a few types (type, various integers, various template and function pointer/reference shapes, …)

That's pre-C++11. All [literal types](https://en.cppreference.com/w/cpp/named_req/LiteralType) are now distinct during the compile-time of C++TMP. The following bullet in the CppReference page on literal types is such a game-changer and my favourite:

" ... * possibly cv-qualified class type that has all of the following properties: ... has a trivial(until C++20)constexpr(since C++20) [destructor](https://en.cppreference.com/w/cpp/language/destructor), ... "

> and the only way to do anything resembling generic programming with it (e.g. treat lists of integers the same way as lists of types) is to erase these distinctions behind type wrappers, which basically defers all type checking until TMP runtime (i.e. when templates are instantiated).

Again, that's pre-C++20. Concepts are now the static semantics of C++TMP. See various examples on that in my teaching where Concepts stop code from compilation way before instantiation.

> there's almost nothing about the type system of C++ TMP that I'd want to replicate in Hylo.

I'm not suggesting that you should. The entire reason why I jumped at this thread was that dependent typing is not exclusively practised for functional correctness. There are various efficiency reasons too. And, Hylo heading for systems programming, that seems quite relevant. Again, I still know nothing about Hylo.

**Reply by dabrahams** [reply]
>> Template expansion makes definition-site checking impractical

> That's pre-C++20. C++ Concepts were such a game-changer. Modern C++TMP is only lenient to that but not such itself.

I'm sorry, but you're misinformed.  C++20 concepts don't support definition-site checking, and are unlikely to do so ever.  It is [not rocket science](https://youtu.be/GsxYnEAZoNI?si=ehLY_xxlajVa5Z2G&t=1034) to get definition-site checking, but if you insist on using “valid expressions” to express constraints instead of doing it the way Haskell, Swift, Rust, Scala, etc. have done it, especially with rampant implicit conversions, unrestricted template specializations, and overloads in the language, well… nobody knows how to solve that problem.

> In theory, a dependently-typed language can also be dynamically typed. (Is there such a language though?) Those are orthogonal topics… A language is dependently-typed as soon as types and (otherwise) values can be passed around alike

I really don't know what to say about this.  If that is true, there's nothing of interest about dependent type systems to discuss in the context of Hylo, because Hylo will of course be dependently typed, as are Swift, Rust, and many other languages I know of.

>> [The static type system of C++TMP] is incredibly weak, having only a few types (type, various integers, various template and function pointer/reference shapes, …)

> That's pre-C++11. 

I think I was well aware of what was in C++11 before I stopped paying attention, although things have expanded since then.

>> and the only way to do anything resembling generic programming with it (e.g. treat lists of integers the same way as lists of types) is to erase these distinctions behind type wrappers, which basically defers all type checking until TMP runtime (i.e. when templates are instantiated).

> Again, that's pre-C++20. 

So you can now write a single template, say `list<...>` and make both `list<int, long, char>` and `list<1, 2, 3>` legal?  That is the problem I'm talking about.  Once you write `list<int_t<1>, int_t<2>, int_t<3>>` you have erased the distinctions, and verifying that your metafunction notionally mapping from *i* to *i + 1* is actually getting numbers as inputs must wait for template instantiation.

> Concepts are now the static semantics of C++TMP. See various examples on that in my teaching where Concepts stop code from compilation way before instantiation.

Well, that's an interesting claim but I don't think I'm ready to dig through all your videos to find the evidence.  Maybe you could show some examples here, and demonstrate how they hold up in the face of unrestricted template specialization and overloading?

**Reply by dabrahams** [reply]
From [cppreference](https://en.cppreference.com/w/cpp/language/constraints):

> Violations of constraints are detected … early **in the template instantiation process**, 

(emphasis mine).  It then goes on to claim:

> which leads to easy to follow error messages:

But it uses an example from non-generic code to show how the messages are easy to follow.

```
std::list<int> l = {3, -1, 10};
std::sort(l.begin(), l.end()); 
```

As soon as `int` becomes `T` in some generic context, the whole thing falls apart, because the generic context can't be checked at its point of definition.  These constraints do nothing to benefit the generic programmer, or even the casual programmer who wants to use a little genericity.

**Reply by hosseinhaeri** [reply]
@dabrahams 

> I'm sorry, but you're misinformed. C++20 concepts don't support definition-site checking, and are unlikely to do so ever. It is [not rocket science](https://youtu.be/GsxYnEAZoNI?si=ehLY_xxlajVa5Z2G&t=1034)

Thanks a lot for the link. I now see what you're after. You don't like two-phase lookup. And, I do confirm I was wrong.

> to get definition-site checking, but if you insist on using “valid expressions” to express constraints instead of doing it the way Haskell, Swift, Rust, Scala, etc. have done it, especially with rampant implicit conversions, unrestricted template specializations, and overloads in the language, well… nobody knows how to solve that problem.

But, I still don't see what the "valid expression" has to do with this. (I see you also invite people to get to you after the talk in the recording you linked to.) And, BTW, I do Haskell and Scala; but, not the other two. I don't see how the "valid expression" is related here.

**Reply by dabrahams** [reply]
Technically, two-phase lookup is just about name resolution.  I'm talking about the fact that type checking of generics is deferred until instantiation time, which is not the same thing  (don't believe everything you read on [cppreference](https://en.cppreference.com/w/cpp/language/two-phase_lookup)).  That delay exists even if names can be resolved in phase 1 because of template specialization and overloads. You can't know what `X<T>` or `f(someT)` means until you know what `T` is, even if you know exactly what `X` and `f` are in phase 1.

See [this paper](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2005/n1758.pdf) (§3.2.3) for the problems with the "valid expression" approach to expressing constraints for separate checking.  Further, If you think through the consequences, you'll see that _even if_ someone could figure out how to implement them, checked valid expression constraints would be incredibly difficult to use in practice.  You'd end up expressing what pseudo-signatures say,  e.g. that the result type of `==` is exactly `bool`, because otherwise there are too many ways to interfere with the compiler's ability to prove the type-soundness of an expression involving `==` in generic context… and that kind of constraint is _far_ more verbose to express with valid expressions than it is with pseudo-signatures.

**Reply by hosseinhaeri** [reply]
> Technically, two-phase lookup is just about name resolution. I'm talking about the fact that type checking of generics is deferred until instantiation time, which is not the same thing (don't believe everything you read on [cppreference](https://en.cppreference.com/w/cpp/language/two-phase_lookup)).

Sorry. Meant two-phase "translation" as opposed to just the "lookup."

> See [this paper](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2005/n1758.pdf) (§3.2.3) for the problems with the "valid expression" approach to expressing constraints for separate checking.

Thanks for the link. I shall take a look later. (It's 01:26 here in the morning. Too late to start reading now even if I had nothing else to do.)

**Reply by dabrahams** [reply]
No, I'm not opposed to multi-phase translation either. Almost every interesting statically-typed language does that. 

This is specifically about having entities that can effectively contain type errors, but can't be type checked. That's what a template is: it can't be type-checked; only its instantiations can.

**Reply by hosseinhaeri** [reply]
Then, educate me please: Which other languages provide multi-phase translation (of templates)? And, what's different between their different phases? Essentially, why would they do in multiple phases?

**Reply by dabrahams** [reply]
No other languages have templates AFAIK. But there are often many phases of translation. In some languages these are all separate phases of translation: parsing, name lookup, type checking, borrow checking, lowering to IR, lowering IR to machine code.

**Reply by hosseinhaeri** [reply]
OK. So, the term you're after is multi-phase compilation -- which is standard compiler engineering terminology. In my sentence below...

> > Meant two-phase "translation" as opposed to just the "lookup."

... however, I'm referring to the specific take of C++ on it for templates. I guess that's the [C++ Templates](https://www.amazon.co.uk/Templates-Complete-Guide-David-Vandevoorde/dp/0321714121/ref=sr_1_1?crid=2CJC3HTREYGO5&keywords=C%2B%2B+Templates&qid=1700616837&sprefix=c%2B%2B+templates%2Caps%2C140&sr=8-1) terminology (Section 1.1.3). And, I seem to understand this is what you don't like in Hylo.

**Reply by dabrahams** [reply]
> OK. So, the term you're after is multi-phase compilation

No thanks, I meant what I said. People have been using that term in compilers for decades. https://learn.microsoft.com/en-us/cpp/preprocessor/phases-of-translation?view=msvc-170 for example

**Reply by hosseinhaeri** [reply]
Sorry @dabrahams for it taking me a few days to come back to you. I was detained elsewhere.

> the only way to do anything resembling generic programming with it (e.g. **treat lists of integers the same way as lists of types**)

(The boldfacing is mine.) That property you're after is what is colloquially called "universal polymorphism" in the community of programming languages. This has been a source of attention since [Stretchy's seminal work](http://cryptyc.cs.depaul.edu/jriely/447/assets/articles/strachey-fundamental-concepts-in-programming-languages.pdf) back in 1967. What confuses me is that universal polymorphism seems to somehow relate in your mind (and in Hylo?) to the phase of type-checking templates, as you put it here:

> This is specifically about having entities that can effectively contain type errors, but can't be type checked. That's what a template is: it can't be type-checked; only its instantiations can.

For your information, both C++ Concepts and Haskell Type-Classes implement "ad-hoc polymorphism" (as opposed to universal polymorphism). But, C++ templates have the property you outline above; whereas type-checking of type-parameterised code in Haskell is performed at the definition site (which is apparently what you like in Hylo). So, I fail to see the relationship between universality and definition-site type-checking. Would you mind explaining the relationship please? (See [this work](https://www.cs.ox.ac.uk/jeremy.gibbons/publications/dgp.pdf) of Jeremy Gibbons with few hits on the ad-hocness of both C++ Concepts and Haskell Type-Classes.)

The other thing which I am unclear about is the reason why you consider template instantiation time the runtime of C++TMP.

> ... which basically defers all type checking until TMP runtime (i.e. when templates are instantiated)

Template instantiation is nothing but a function application during compile-time of C++. But, function application per se needs not to take place during the runtime of a language. For example, in ordinary C++, `constexpr` functions (when passed `constexpr` arguments) and `consteval` functions (unconditionally) get invoked too but during compile-time.

So, even though -- contrary to my previous words -- I admit that concept enforcement takes place only after instantiation, the following quote from [cppreference](https://en.cppreference.com/w/cpp/language/constraints) that you yourself brought up...

> Violations of constraints are detected at compile time, **early** in the template instantiation process

... suggests that it makes perfect sense to take that **early** phase the compile-time of C++TMP, and, hence, C++ Concepts the static semantics of C++TMP.

**Reply by dabrahams** [reply]
I only raised lack of polymorphism as an example of the [weakness of TMP's static type system](https://github.com/orgs/hylo-lang/discussions/1126#discussioncomment-7621142).  I consider instantiation time the runtime of TMP because it's when the compile-time computation, e.g. of new types, happens.  `constexpr` and `consteval` stuff is not TMP—it doesn't even necessarily involve templates—but unlike template metaprograms, definitions marked `constexpr` and `consteval` _are_ separately typechecked, independent of whether they are used anywhere…as long as those definitions are not templates ;-)

> suggests that it makes perfect sense to take that early phase the compile-time of C++TMP, and, hence, C++ Concepts the static semantics of C++TMP.

Sorry to be blunt, but that is just wrong. Template instantiation, and thus concept checks, take place during the _evaluation_ of (compile-time) expressions.  Static type checking is something we apply to definitions _without having to evaluate expressions_ depending on those definitions.  Dynamic type checking is what happens when type soundness is validated as a consequence of the evaluation of expressions.  To the extent that adding C++ concept constraints to templates makes any checking happen earlier, it is the compile-time equivalent of the assertion in this Python:

```
x = [1, 2, 3, 4]
def inner(i):
  return x[i]
  
def outer(i):
  assert type(i) is int, "concept check failed"
  return inner(i)
  
print(outer("foo"))
```

The program would fail without this assertion, but we'd get a longer error backtrace from Python.  This backtrace corresponds exactly to a template instantiation backtrace in C++, which can be shortened by adding concept checks on outer templates.

**Reply by emdash** [reply]
I have a couple larger examples that might be more interesting. I'm happy to elucidate how they work (insofar as I understand it) for those who don't read the notation.

## Dates

Here's an example which captured my attention early: https://github.com/emdash/idris-datetime/blob/main/src/Date.idr

The heart of it is a dependent record for dates which makes it impossible to represent an invalid date, like 2023-2-29, or 2024-11-31:

```haskell
export
record Date where
  constructor MkDate
  year:  Nat
  month: Month
  day:   Month.Day (IsLeapYear year) month
```

The rest of the library does involve the kind of "jumping through hoops" to satisfy predicates that I suspect Hylo devs want to avoid, but I chose to go through this exercise. This shows how dependent types can get a bit ugly.

## Dimensional Analysis

https://github.com/emdash/ampii/blob/main/src/Measures.idr

This is a toy dimensional analysis system. It's not maximally generic. It doesn't handle all the fundamental units. But I think it shows how pleasant dependent types can be.

**Reply by dabrahams** [reply]
We are well aware of these examples. The use of templates for dimensional analysis goes back to Barton & Nackman, well before standardization. https://www.amazon.com/Scientific-Engineering-Introduction-Advanced-Techniques/dp/0201533936?dplnkId=da1334c3-16c6-4100-b36e-75e68e580bc2&nodl=1

We intend to cover these use cases; The features required are not what we have concerns about when it comes to "dependent type system support"

**Comment by waydan** [comment]
I've been wondering after reading this: is there something dependent types gets you that isn't available in design by contract? AFAICT, expressing that a value cannot be greater than the input is a post-condition. I don't have experience with languages like idris2, but I've seen dependent types used to constrain valid input which looks a lot like preconditions.

**Reply by dabrahams** [reply]
🍏 🍎 and 🍊 🍊. DbC is a methodology and dependent types are a language feature. That feature, like basically all features of static type systems, lets you prove at compile time that certain contracts are upheld. Most features called "DbC support" check contracts at runtime and have no role in the type system. 

Then again. I don't see how you could possibly validate input using a (static) type system feature. 

HTH

**Reply by hosseinhaeri** [reply]
DbC is essentially not concerned about the stage at which the contract enforcement takes place. Dependent typing is strictly concerned about compile-time.

More to the point: Not everything in dependent typing is about correctness (by contract). A whole world of optimisation only surfaces in the presence of such expressiveness. Expression templates are the closest that C++TMP gets to wrt optimisation, for example. The gist of it is that, when your compiler knows more about your code at compile-time, it can optimise things much better. And, my little research about Hylo is that it's a systems programming language. Such languages are expected to be speed-thirsty. You'd want to jump on such optmisations, I guess.

**Comment by emdash** [comment]
Hot take: Numbers in types are the "gateway drug" of dependent typing. Once I saw the value of this, I wanted more.

For me, Frankly, it was C++ templates that sent me down the rabbit hole of dependent types. It was C++ that put the concept in my head, but for now I'm quite keen on learning Idris.

I think anyone who's spent any serious time with modern C++ has started to form some intuitive notion of dependent types. The trouble with pointing to C++ as an example of *anything* is that C++ is, to put it delicately, a *legacy language*. So it's not the clearest expression of the idea.

> Template instantiation, and thus concept checks, take place during the _evaluation_ of (compile-time) expressions. Static type checking is something we apply to definitions _without having to evaluate expressions_ depending on those definitions. Dynamic type checking is what happens when type soundness is validated as a consequence of the evaluation of expressions. To the extent that adding C++ concept constraints to templates makes any checking happen earlier, it is the compile-time equivalent of the assertion in this Python:
> 
> ```
> x = [1, 2, 3, 4]
> def inner(i):
>   return x[i]
>   
> def outer(i):
>   assert type(i) is int, "concept check failed"
>   return inner(i)
>   
> print(outer("foo"))
> ```
> 
> The program would fail without this assertion, but we'd get a longer error backtrace from Python. This backtrace corresponds exactly to a template instantiation backtrace in C++, which can be shortened by adding concept checks on outer templates.

Uh. What? I checked, and *assertions* throw an exceptions at *runtime*.

C++ template instantiation may be "morally dynamic", but template instantiation still happens at *compile time*.  That's where this discussion of "phases" comes in. For my part, as a developer and not a compiler author, I don't really care whether the error is found by the compiler itself, or a constraint supplied by a library author, so long as it *finds* it.

My interset is more about giving library authors the tools needed to produce good libraries, rather than proving everything reduces to lambda terms.

**Reply by kyouko-taiga** [reply]
Though this conversation is interesting, I am afraid this thread has moved away from its original topic.

We should not be debating the pros and cons of declaration-site checking (e.g., Swift generics) vs use-site checking (e.g., C++ templates) here. I'm happy to explain why we strongly believe the latter is bad and why the end user actually cares, but we should do that in another thread. Anyone feel free to ask the question if you're interested.

**Reply by dabrahams** [reply]
I have lots to say about that too. But I agree, not in this thread please.  

Also, @emdash, I'm going choose to believe  you know nothing at all about the history and ongoing professional missions of those you are chatting with; otherwise, I think (or hope anyway) nothing like the implication of that closing sentence would have made it into this thread. Giving library authors the tools needed to produce good libraries is at the heart of the design choices in question.

**Reply by emdash** [reply]
Fair enough, I do not.

**Comment by dabrahams** [comment]
Thanks for your participation everyone.  I think we failed to set the context for this discussion properly, which resulted in a lot of static, so I am closing it in favor #1200.  I hope this helps!
